tests/perf/shared_token_bucket.cc

```diff
@@ -0,0 +1,295 @@
+            [this] {
+                auto p = std::chrono::duration<double>(rest(testing::local_random_engine));
+                return seastar::sleep(std::chrono::duration_cast<std::chrono::microseconds>(p)).then([this] {
+                    _iterations++;
+                    auto until = clock_type::now() + std::chrono::duration<double>(busy(testing::local_random_engine));
+                    do {
+                    } while (clock_type::now() < until && keep_going);
+                });
+            }
+        );
+    }
+
+    future<> terminate() {
+        assert(stopped.has_value());
+        keep_going = false;
+        auto f = std::move(*stopped);
+        stopped.reset();
+        return f.finally([this] {
+            fmt::print("{}-hog {} iters\n", this_shard_id(), std::exchange(_iterations, 0));
+        });
+    }
+};
+
+template <typename TokenBucket>
+struct context {
+    using worker_t = worker<TokenBucket>;
+    TokenBucket tb;
+    seastar::sharded<worker_t> w;
+    seastar::sharded<hog> h;
+
+    static constexpr uint64_t rate = 1000000;
+    static constexpr uint64_t limit = rate / 2000;
+    static constexpr uint64_t threshold = 1;
+
+    context() : tb(rate, limit, threshold)
+    {
+        w.start(std::ref(tb)).get();
+        h.start(std::chrono::microseconds(300), std::chrono::microseconds(100)).get();
+        fmt::print("Created tb {}t/s (limit {} threshold {})\n", tb.rate(), tb.limit(), tb.threshold());
+    }
+
+    ~context() {
+        h.stop().get();
+        w.stop().get();
+    }
+
+    template <typename Fn>
+    future<> run_workers(Fn&& fn) {
+        auto start = clock_type::now();
+        return w.map_reduce0(std::forward<Fn>(fn), statistics{}, accumulate).then([this, start] (statistics st) {
+            auto delay = std::chrono::duration_cast<std::chrono::duration<double>>(clock_type::now() - start).count();
+            fmt::print("effective rate is {:.1f}t/s, [{} ... {}]\n", st.total / delay, st.min, st.max);
+        });
+    }
+
+    future<> test_sleeping() {
+        fmt::print("---8<---\n");
+        return run_workers(&worker_t::work_sleeping);
+    }
+
+    future<> test_yielding() {
+        fmt::print("---8<---\n");
+        return run_workers(&worker_t::work_yielding);
+    }
+
+    future<> test_sleeping_with_hog() {
+        fmt::print("---8<---\n");
+        return h.invoke_on_all(&hog::work).then([this] {
+            return run_workers(&worker_t::work_sleeping).then([this] {
+                return h.invoke_on_all(&hog::terminate);
+            });
+        });
+    }
+};
+
+struct perf_capped_context : public context<capped_token_bucket_t> {};
+struct perf_pure_context : public context<pure_token_bucket_t> {};
+
+PERF_TEST_F(perf_capped_context, sleeping_throughput) { return test_sleeping(); }
+PERF_TEST_F(perf_capped_context, yielding_throughput) { return test_yielding(); }
+PERF_TEST_F(perf_capped_context, sleeping_throughput_with_hog) { return test_sleeping_with_hog(); }
```

Comment:
Please add lots of comments explaining the test scenario and documenting what a good result is (and why). I'm sufficiently detached from the internals that it's hard for me to figure it out directly.

Reply:
Yup, my bad. Added